lib/compress/zstd_compress.c

```diff
@@ -2378,7 +2438,7 @@ ZSTD_entropyCompressSequences(seqStore_t* seqStorePtr,
                              void* dst, size_t dstCapacity,
                              size_t srcSize,
                              void* entropyWorkspace, size_t entropyWkspSize,
-                             int bmi2)
+                             int bmi2, U32 const canEmitUncompressed)
```

Comment:
Interesting, 
so you want to be able to forbid `ZSTD_entropyCompressSequences()`
from sending non-compressed sequences
therefore including cases where it's not advantageous.

1) Is that just a test feature, or a permanent one ?
2) If permanent, could you explain me in which cases this control is necessary ?
3) the specification doesn't require the compressed data to be smaller than uncompressed one. _But_ a compressed block must still be smaller than a full block (a.k.a 128 KB for the general case). The full block also includes the literals block. What would happen in this case ?

Reply:
So this is a permanent feature - the reason I include this here is that I want to avoid sending uncompressed/RLE blocks when it would destroy repcode history in a bad way while block splitting. i.e., if the next block's first three sequences contain a repcode, we shouldn't try to emit an uncompressed or RLE block, since then the next block's repcode sequence will reference a sequence that is now "lost" as part of an uncompressed or RLE block and therefore invalid.

I suppose that this case, it could be possible that we emit a "compressed" block that is larger than an uncompressed one - basically the a similar limitation as superblocks, though I'm not quite sure how we could get around this while block splitting.

Reply:
I still have one question on this point : 

Can you guarantee that this mechanism (forcing the output to be compressed even when disadvantageous) is not going to produce a block size large than a full block ?
If yes, can you add an `assert()` ?

Reply:
I don't think I can guarantee that, at least from this particular codepath.

One thing we could potentially do is: if it turns out that our split, compressed block's total size is larger than 128 KB, then what we could do is trigger a re-compression using the traditional method. Do you think that'd be sufficient? Is it accurate to say that the upper bound on a compressed block is then 131075 bytes? (128K + 3 byte block header).

Reply:
> Is it accurate to say that the upper bound on a compressed block is then 131075 bytes? (128K + 3 byte block header).

Well, depending on where the size is measured, yes this can be correct.

Not though that there are several places in the code which do not include the block header cost. 

> One thing we could potentially do is: if it turns out that our split, compressed block's total size is larger than 128 KB, then what we could do is trigger a re-compression using the traditional method (single block). Do you think that'd be sufficient?

Yes.

That being said, the limitation is lower than that :
no _individual_ block can be `> 128 KB`.

If the original block is split, and the sum of both parts is > 128 KB, this is not a "problem", in the sense that it's still valid (inefficient sure, but valid).
But if **one** block is > 128 KB, then it's beyond spec, and can break the decoder.
That's what we want to avoid.

Reply:
Okay good to know - I believe in this case we'd need to go with this stricter limitation of not allowing the split blocks cumulatively to be over 128K (and in so doing prevent any single one from being over 128K).

This is because the only case in which we'd potentially emit a compressed block larger than 128K is when we forcibly disable uncompressed or RLE blocks because doing so would erroneously destroy the repcode history. So in this situation the following two statements are simultaneously true: 1) emitting a compressed block instead of uncompressed/RLE block is necessary to maintain repcode history 2) emitting a compressed block may exceed the 128K limit.

Thus, we have to basically re-compress in case statement 2) turns out to be true. The latest commit adds in this functionality.